### ISSUE TITLE
[jvm-package]  JVM crash casued by getting the leaf indexes of a large test data set from gbtree model

### ISSUE THREAD

**Issue by MScracker** [dmlc/xgboost#2088]
## Environment info
Operating System: linux 
Package used (python/R/jvm/C++):  jvm(spark)
`xgboost` version used: 0.6

```
    import DataUtils._
    val testRDDLP = testAssembleDF.select("features", "label").rdd
      .map {
        case Row(feature: Vector, label: Int) => LabeledPoint(label, feature)
      }
    testRDDLP.cache() 
    val testSetDMatrix= new DMatrix(testRDDLP.toLocalIterator)
    val testLeafIndex = xgboostClassificationModel.booster.predictLeaf(testSetDMatrix)
```
I want to get the leaf index of test set (100 milllion + records) by converting  RDD to DMatrix using **toLocalIterator** method, but I always got the same error: A fatal error has been detected by the Java Runtime Environment, then JVM crashed.

I tried to convert RDD to DMatrix using map:
   ```
 val leafIndexRDD = testRDDLP.mapPartitions { samples =>
      var leafIndex: Array[Array[Float]] = null
      if (samples.hasNext) {
        val dMatrix = new DMatrix(samples)
        leafIndex = xgboostClassificationModel.booster.predictLeaf(dMatrix)
      } 
      Iterator(leafIndex)
    }.cache()
```
but  date type of leafIndexRDD is **RDD[Array[Array[Float]]]**, which i wish is Dataframe.  so i tried many times converting  RDD[Array[Array[Float]]] to Dataframe, but failed.  My question is that **how to get the leaf index of a large test data set** and  the ideal data type is Dataframe that  I can use in the following model training.
